sops.go

```diff
@@ -0,0 +1,259 @@
+package sops
+
+import (
+	"crypto/sha512"
+	"fmt"
+	"go.mozilla.org/sops/aes"
+	"go.mozilla.org/sops/kms"
+	"go.mozilla.org/sops/pgp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const DateFormat = "2006-01-02T15:04:05Z"
+
+const DefaultUnencryptedSuffix = "_unencrypted"
+
+type Error string
+
+func (e Error) Error() string { return string(e) }
+
+const MacMismatch = Error("MAC mismatch")
+
+type TreeItem struct {
+	Key   string
+	Value interface{}
+}
+
+type TreeBranch []TreeItem
+
+type Tree struct {
+	Branch   TreeBranch
+	Metadata Metadata
+}
+
+func (tree TreeBranch) WalkValue(in interface{}, path []string, onLeaves func(in interface{}, path []string) (interface{}, error)) (interface{}, error) {
+	switch in := in.(type) {
+	case string:
+		return onLeaves(in, path)
+	case int:
+		return onLeaves(in, path)
+	case bool:
+		return onLeaves(in, path)
+	case TreeBranch:
+		return tree.WalkBranch(in, path, onLeaves)
+	case []interface{}:
+		return tree.WalkSlice(in, path, onLeaves)
+	default:
+		return nil, fmt.Errorf("Cannot walk value, unknown type: %T", in)
+	}
+}
+
+func (tree TreeBranch) WalkSlice(in []interface{}, path []string, onLeaves func(in interface{}, path []string) (interface{}, error)) ([]interface{}, error) {
+	for i, v := range in {
+		newV, err := tree.WalkValue(v, path, onLeaves)
+		if err != nil {
+			return nil, err
+		}
+		in[i] = newV
+	}
+	return in, nil
+}
+
+func (tree TreeBranch) WalkBranch(in TreeBranch, path []string, onLeaves func(in interface{}, path []string) (interface{}, error)) (TreeBranch, error) {
+	for i, item := range in {
+		newV, err := tree.WalkValue(item.Value, append(path, item.Key), onLeaves)
+		if err != nil {
+			return nil, err
+		}
+		in[i].Value = newV
+	}
+	return in, nil
+}
+
+func (tree Tree) Encrypt(key string) (string, error) {
+	hash := sha512.New()
+	_, err := tree.Branch.WalkBranch(tree.Branch, make([]string, 0), func(in interface{}, path []string) (interface{}, error) {
+		bytes, err := toBytes(in)
+		if !strings.HasSuffix(path[len(path)-1], tree.Metadata.UnencryptedSuffix) {
+			var err error
+			in, err = aes.Encrypt(in, key, []byte(strings.Join(path, ":")+":"))
+			if err != nil {
+				return nil, fmt.Errorf("Could not encrypt value: %s", err)
+			}
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Could not convert %s to bytes: %s", in, err)
+		}
+		hash.Write(bytes)
+		return in, err
+	})
+	if err != nil {
+		return "", fmt.Errorf("Error walking tree: %s", err)
+	}
+	return fmt.Sprintf("%X", hash.Sum(nil)), nil
+}
+
+func (tree Tree) Decrypt(key string) (string, error) {
+	hash := sha512.New()
+	_, err := tree.Branch.WalkBranch(tree.Branch, make([]string, 0), func(in interface{}, path []string) (interface{}, error) {
+		var v interface{}
+		if !strings.HasSuffix(path[len(path)-1], tree.Metadata.UnencryptedSuffix) {
+			var err error
```

Comment:
It's cleaner to predefine your variables at the top, right under the function name:
```go
func (tree Tree) Decrypt(key string) (string, error) {
var (
        v interface{}
        err error
)
```